src/v/archival/archiver_operations_impl.cc

```diff
@@ -327,9 +327,117 @@ class archiver_operations_impl : public archiver_operations_api {
 
     /// Add metadata to the manifest by replicating archival metadata
     /// configuration batch
-    ss::future<result<admit_uploads_result>>
-    admit_uploads(retry_chain_node&, upload_results_list) noexcept override {
-        co_return error_outcome::unexpected_failure;
+    ss::future<result<admit_uploads_result>> admit_uploads(
+      retry_chain_node& workflow_rtc,
+      upload_results_list upl_res) noexcept override {
+        // Generate archival metadata for every uploaded segment and
+        // apply it in offset order.
+        // Stop on first error.
+        // Validate consistency.
+        try {
+            size_t num_segments = upl_res.results.size();
+            if (
+              num_segments != upl_res.stats.size()
+              || num_segments != upl_res.metadata.size()) {
+                vlog(
+                  _rtclog.error,
+                  "Bad 'admit_uploads' input. Number of segments: {}, number "
+                  "of "
+                  "stats: {}, number of metadata records: {}",
+                  num_segments,
+                  upl_res.stats.size(),
+                  upl_res.metadata.size());
+                co_return error_outcome::unexpected_failure;
+            }
+            bool validation_required
+              = !config::shard_local_cfg()
+                   .cloud_storage_disable_upload_consistency_checks();
+            auto part = _pm->get_partition(upl_res.ntp);
+            std::vector<cloud_storage::segment_meta> metadata;
+            for (size_t ix = 0; ix < num_segments; ix++) {
+                auto sg = upl_res.results.at(ix);
+                auto st = upl_res.stats.at(ix);
+                auto meta = upl_res.metadata.at(ix);
+
+                if (sg != upload_result::success) {
+                    break;
+                }
+                // validate meta against the record stats
+                if (st.has_value() && validation_required) {
+                    if (!this->segment_meta_matches_stats(
+                          meta, st.value(), _rtclog)) {
+                        break;
+                    }
+                } else {
```

Comment:
Are there any expectations here with regards to the order of the segment metadata? Based on how it's written we're expecting that the segments are ordered by offset, and that breaking here means that we only admit the first contiguous set that has succeeded, is that right?